### PR TITLE
Add Background Blur and Background Replacement Processors

### PR DESCRIPTION
**Pull request by hekmatba**
### Description of changes:
- Add Background Blur and Background Replacement processors.
- Updated demo app to support background filters.

#### Manual test cases (add more as needed):

- [ ] Join meeting without CallKit
- [ ] Join meeting with CallKit as Incoming
- [ ] Join meeting with CallKit as Outgoing
- [ ] Leave meeting
- [ ] Rejoin meeting
- [ ] See metrics
- [ ] See attendee presence
- [ ] Send audio
- [ ] Receive audio
- [ ] Mute self
- [ ] Unmute self
- [ ] See local mute indicator when muted
- [ ] See remote mute indicator when other is muted
- [ ] Enable and disable local video
- [ ] Enable and disable remote video from remote side
- [ ] Send local video
- [ ] Pause and resume remote video
- [ ] Switch local camera
- [ ] Receive remote screen share

#### Integration validation (required before release)

- [ ] Unit tests pass
- [ ] Build SDK against simulator with release options
- [ ] Build SDK against device with release options
- [ ] Build SDK against simulator with debug options
- [ ] Build SDK against device with debug options
- [ ] Test Demo against simulator with SDK (debug build) in workspace
- [ ] Test Demo against device with SDK (debug build) in workspace
- [ ] Test DemoObjC against simulator with SDK (debug build) in workspace
- [ ] Test DemoObjC against device with SDK (debug build) in workspace
- [ ] Test Demo against simulator with SDK.framework (release build)
- [ ] Test Demo against device with SDK.framework (release build)
- [ ] Test DemoObjC against simulator with SDK.framework (release build)
- [ ] Test DemoObjC against device with SDK.framework (release build)

### Screenshots, if available:

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: .github/workflows/codecov.yml [hunk 1]
```diff
@@ -37,6 +37,20 @@ jobs:
           tar -xzf AmazonChimeSDKMedia.tar.gz
           cp -R ./AmazonChimeSDKMedia.framework ./AmazonChimeSDK
 
+      # Download amazon-chime-sdk-machine-learning binary from AWS S3
```
Comment (ziyiz-amzn) [inline .github/workflows/codecov.yml:40]: This change should have been merged by previous PR.

File: AmazonChimeSDK/AmazonChimeSDK.xcodeproj/project.pbxproj [hunk 2]
```diff
@@ -1928,6 +2022,7 @@
 				OTHER_LDFLAGS = (
 					"-ObjC",
 					"$(inherited)",
+					"-lc++\n-lc++",
```
Comment (ziyiz-amzn) [inline AmazonChimeSDK/AmazonChimeSDK.xcodeproj/project.pbxproj:2025]: one will be fine.

File: .github/workflows/codecov.yml [hunk 3]
```diff
@@ -37,6 +37,20 @@ jobs:
...
+          aws configure set aws_access_key_id ${{ secrets.AWS_ACCESS_KEY_ID }} --profile jenkins-automated-test
+          aws configure set aws_secret_access_key ${{ secrets.AWS_SECRET_ACCESS_KEY }} --profile jenkins-automated-test
+          aws \
+            --profile jenkins-automated-test \
+            s3api get-object \
+            --bucket amazon-chime-sdk-ios-internal \
+            --key master/machine-learning/latest/AmazonChimeSDKMachineLearning.tar.gz \
+            AmazonChimeSDKMachineLearning.tar.gz
+          tar -xzf AmazonChimeSDKMachineLearning.tar.gz
+          cp -R ./AmazonChimeSDKMachineLearning.framework ./AmazonChimeSDK
```
[the remark this answers is not in the document]
Comment (hekmatba) [reply .github/workflows/codecov.yml:52]: It looks like the .framework is also there. Switching to .xcframework

File: AmazonChimeSDK/AmazonChimeSDK.xcodeproj/project.pbxproj [hunk 4]
```diff
@@ -256,6 +255,21 @@
...
+		CFA2A1542807DAAA008FDFF5 /* BackgroundFilter.swift in Sources */ = {isa = PBXBuildFile; fileRef = CFA2A1532807DAAA008FDFF5 /* BackgroundFilter.swift */; };
+		CFA2A16328085D23008FDFF5 /* BackgroundBlurStrength.swift in Sources */ = {isa = PBXBuildFile; fileRef = CFA2A16228085D23008FDFF5 /* BackgroundBlurStrength.swift */; };
+		CFA2A17028087AE8008FDFF5 /* BackgroundFilterTest.swift in Sources */ = {isa = PBXBuildFile; fileRef = CFA2A16F28087AE8008FDFF5 /* BackgroundFilterTest.swift */; };
+		CFA2A17728091499008FDFF5 /* AmazonChimeSDKMedia.framework in Frameworks */ = {isa = PBXBuildFile; fileRef = 5ADEAFA024293B2700801EA5 /* AmazonChimeSDKMedia.framework */; };
+		CFD14AAC281B4A8300C8D8CA /* selfie_segmentation_landscape.tflite in Resources */ = {isa = PBXBuildFile; fileRef = CFD14AAB281B4A8300C8D8CA /* selfie_segmentation_landscape.tflite */; };
+		CFD14AAD281B4A8300C8D8CA /* selfie_segmentation_landscape.tflite in Resources */ = {isa = PBXBuildFile; fileRef = CFD14AAB281B4A8300C8D8CA /* selfie_segmentation_landscape.tflite */; };
```
[the remark this answers is not in the document]
Comment (hekmatba) [reply AmazonChimeSDK/AmazonChimeSDK.xcodeproj/project.pbxproj:271]: I think one for AmazonChimeSDK and other for AmazonChimeSDKTests targets

File: AmazonChimeSDK/AmazonChimeSDK/audiovideo/video/backgroundfilter/BackgroundFilterProcessor.swift [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+    /// Creates the alpha mask [0-255] of the foreground image using `SegmentationProcessor`.
+    ///
+    /// - Parameters:
+    ///   - inputFrameCG: Input CGImage frame to produce the foreground image.
+    ///   - inputFrameCI: Input CIImage frame to produce the foreground image.
+    ///
+    /// - Returns: Alpha mask CGImage of the foreground.
+    public func createForegroundAlphaMask(inputFrameCG: CGImage,
+                                          inputFrameCI: CIImage) -> CIImage? {
+        #if canImport(AmazonChimeSDKMachineLearning)
```
Comment (hokyungh) [inline AmazonChimeSDK/AmazonChimeSDK/audiovideo/video/backgroundfilter/BackgroundFilterProcessor.swift:64]: nit: Can we do

```
#if !canImport(AmazonChimeSDKMachineLearning)
   return
#endif
```

?

File: AmazonChimeSDK/AmazonChimeSDK/audiovideo/video/backgroundfilter/BackgroundFilterProcessor.swift [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+        // Number of the input image color space channels.
+        let imageChannels = inputFrameCG.bitsPerPixel / inputFrameCG.bitsPerComponent
+
+        // Update the buffer pool dimensions if the new frame does not match the previous frame dimensions.
+        if bufferPool == nil || inputFrameCG.width != bufferPoolWidth || inputFrameCG.height != bufferPoolHeight {
+
```
Comment (hokyungh) [inline AmazonChimeSDK/AmazonChimeSDK/audiovideo/video/backgroundfilter/BackgroundFilterProcessor.swift:70]: unnecessary space.

File: AmazonChimeSDK/AmazonChimeSDK/audiovideo/video/backgroundfilter/BackgroundFilterProcessor.swift [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+                             "how to import `selfie_segmentation_landscape.tflite` as a bundle resource to your project.")
+                return nil
+            } catch {
+                logger.error(msg: "Unexpected error: \(error).")
+            }
+        }
+
+        // Calculate the scale and aspect ratio factor to downscale.
+        let downSampleScale = Double(segmentationProcessorHeight) / Double(inputFrameCG.height)
+        let downSampleAspectRatio = Double((Double(inputFrameCG.height)/Double(inputFrameCG.width)) /
```
Comment (hokyungh) [inline AmazonChimeSDK/AmazonChimeSDK/audiovideo/video/backgroundfilter/BackgroundFilterProcessor.swift:89]: spacing on `height)/Double(` and `rHeight)/Double`

File: AmazonChimeSDK/AmazonChimeSDK/audiovideo/video/backgroundfilter/BackgroundFilterProcessor.swift [hunk 8]
```diff
@@ -0,0 +1,236 @@
...
+                                           (Double(segmentationProcessorHeight)/Double(segmentationProcessorWidth)))
+        // Down sample the image.
+
```
Comment (hokyungh) [inline AmazonChimeSDK/AmazonChimeSDK/audiovideo/video/backgroundfilter/BackgroundFilterProcessor.swift:92]: extra space

File: AmazonChimeSDK/AmazonChimeSDK/audiovideo/video/backgroundfilter/BackgroundFilterProcessor.swift [hunk 9]
```diff
@@ -0,0 +1,236 @@
...
+        guard let downSampleImage: CIImage = resizeImage(image: inputFrameCI,
+                                                         scale: downSampleScale,
+                                                         aspectRatio: downSampleAspectRatio)
+        else {
+            return nil
+        }
+
+        // Check if segmentation model has loaded.
+        if segmentationProcessor.getModelState() != CwtModelState.LOADED {
```
Comment (hokyungh) [inline AmazonChimeSDK/AmazonChimeSDK/audiovideo/video/backgroundfilter/BackgroundFilterProcessor.swift:101]: shouldn't we have this if on the top before `imageChannels`?

File: AmazonChimeSDK/AmazonChimeSDK/audiovideo/video/backgroundfilter/BackgroundFilterProcessor.swift [hunk 10]
```diff
@@ -0,0 +1,236 @@
...
+        else {
+            logger.error(msg: "Error creating CGImage of the foreground mask.")
+            return nil
+        }
+
+        let maskImageCi = CIImage(cgImage: maskImage)
+
+        // Calculate the scale and aspect ratio factor to upscale.
+        let upSampleScale = Double(inputFrameCG.height) / Double(segmentationProcessorHeight)
+        let upSampleAspectRatio = Double((Double(segmentationProcessorHeight)/Double(segmentationProcessorWidth)) /
```
Comment (hokyungh) [inline AmazonChimeSDK/AmazonChimeSDK/audiovideo/video/backgroundfilter/BackgroundFilterProcessor.swift:146]: hmmm just curious can't we just do `upSampleAspectRatio  = 1 / downSampleAspectRatio`?

File: AmazonChimeSDK/AmazonChimeSDK/audiovideo/video/backgroundfilter/backgroundblur/BackgroundBlurVideoFrameProcessor.swift [hunk 11]
```diff
@@ -0,0 +1,152 @@
...
+
+        // Create the final `CVPixelBuffer` for the output image.
+        var mergedImageBuffer: CVPixelBuffer?
+        guard let bufferPool = backgroundFilterProcessor.getBufferPool() else {
+            backgroundFilterProcessor.logger.error(msg: "Error retrieving final buffer pool.")
+            return
+        }
+        CVPixelBufferPoolCreatePixelBuffer(nil, bufferPool, &mergedImageBuffer)
+        guard let validMergedImageBuffer = mergedImageBuffer else {
+            backgroundFilterProcessor.logger.error(msg: "Error creating validMergedImageBuffer.")
```
Comment (hokyungh) [inline AmazonChimeSDK/AmazonChimeSDK/audiovideo/video/backgroundfilter/backgroundblur/BackgroundBlurVideoFrameProcessor.swift:109]: can we have better log than putting variable name?

File: AmazonChimeSDK/AmazonChimeSDK/audiovideo/video/backgroundfilter/backgroundreplacement/BackgroundReplacementVideoFrameProcessor.swift [hunk 12]
```diff
@@ -0,0 +1,160 @@
...
+
+        // Create the final `CVPixelBuffer` for the output image.
+        var mergedImageBuffer: CVPixelBuffer?
+        guard let bufferPool = backgroundFilterProcessor.getBufferPool() else {
+            backgroundFilterProcessor.logger.error(msg: "Error retrieving final buffer pool.")
+            return
+        }
+        CVPixelBufferPoolCreatePixelBuffer(nil, bufferPool, &mergedImageBuffer)
+        guard let validMergedImageBuffer = mergedImageBuffer else {
+            backgroundFilterProcessor.logger.error(msg: "Error creating validMergedImageBuffer.")
```
Comment (hokyungh) [inline AmazonChimeSDK/AmazonChimeSDK/audiovideo/video/backgroundfilter/backgroundreplacement/BackgroundReplacementVideoFrameProcessor.swift:113]: ditto

File: AmazonChimeSDK/AmazonChimeSDK/internal/utils/ImageConversionUtils.swift [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+        let width = cgImage.width
+
+        let bitsPerComponent = cgImage.bitsPerComponent
+        let bytesPerRow = cgImage.bytesPerRow
+        let size = height * bytesPerRow
+
+        var byteArray: [UInt8] = [UInt8](repeating: 0, count: size)
+
+        guard let contextRef = CGContext(data: &byteArray,
+                                   width: width,
```
Comment (hokyungh) [inline AmazonChimeSDK/AmazonChimeSDK/internal/utils/ImageConversionUtils.swift:78]: spacing.

File: AmazonChimeSDK/AmazonChimeSDK/audiovideo/video/backgroundfilter/BackgroundFilterProcessor.swift (hunk 9 above)
Comment (hekmatba) [reply AmazonChimeSDK/AmazonChimeSDK/audiovideo/video/backgroundfilter/BackgroundFilterProcessor.swift:101]: We need to check if the model was loaded after the initialization on line 74

File: AmazonChimeSDKDemo/AmazonChimeSDKDemo/VideoTileCell.swift [hunk 14]
```diff
@@ -63,7 +63,7 @@ class VideoTileCell: UICollectionViewCell {
         onTileButton.addTarget(self, action: #selector(onTileButtonClicked), for: .touchUpInside)
         onTileButton.isSelected = isVideoPausedByUser
         updateVideoSubscriptionsButton.isHidden = false
-        videoFiltersButton.isHidden = false
+        videoFiltersButton.isHidden = true
```
Comment (ziyiz-amzn) [inline AmazonChimeSDKDemo/AmazonChimeSDKDemo/VideoTileCell.swift:66]: Beside this change, can we hide the option button for local video when ML not imported? The alert controller with a `none` action seems weird.

Screen Shot 2022-05-04 at 8 51 40 PM

**Comment by hekmatba** [comment]
New pull request was opened https://github.com/aws/amazon-chime-sdk-ios/pull/463

**Pull request closed without merging**